CMakeLists.txt

```diff
@@ -47,16 +47,55 @@ set(CMAKE_CXX_EXTENSIONS ON CACHE INTERNAL "")
 set(CMAKE_CXX_SCAN_FOR_MODULES OFF CACHE INTERNAL "")
 set(CMAKE_CXX_VISIBILITY_PRESET hidden)
 
-set(Seastar_TESTING ON CACHE BOOL "" FORCE)
-set(Seastar_API_LEVEL 7 CACHE STRING "" FORCE)
-set(Seastar_DEPRECATED_OSTREAM_FORMATTERS OFF CACHE BOOL "" FORCE)
-set(Seastar_APPS ON CACHE BOOL "" FORCE)
-set(Seastar_EXCLUDE_APPS_FROM_ALL ON CACHE BOOL "" FORCE)
-set(Seastar_EXCLUDE_TESTS_FROM_ALL ON CACHE BOOL "" FORCE)
-set(Seastar_IO_URING OFF CACHE BOOL "" FORCE)
-set(Seastar_SCHEDULING_GROUPS_COUNT 16 CACHE STRING "" FORCE)
-set(Seastar_UNUSED_RESULT_ERROR ON CACHE BOOL "" FORCE)
-add_subdirectory(seastar)
+get_property(is_multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if(is_multi_config)
+    find_package(Seastar)
+    # this is atypical compared to standard ExternalProject usage:
+    # - Seastar's build system should already be configured at this point.
+    # - We maintain separate project variants for each configuration type.
+    #
+    # Benefits of this approach:
+    # - Allows the parent project to consume the compile options exposed by
+    #   .pc file. as the compile options vary from one config to another.
+    # - Allows application of config-specific settings
+    # - Enables building Seastar within the parent project's build system
+    # - Facilitates linking of artifacts with the external project target,
+    #   establishing proper dependencies between them
+    include(ExternalProject)
+
+    ExternalProject_Add(Seastar
+        SOURCE_DIR "${PROJECT_SOURCE_DIR}/seastar"
+        BINARY_DIR "${CMAKE_BINARY_DIR}/$<CONFIG>/seastar"
+        CONFIGURE_COMMAND ""
+        BUILD_COMMAND ${CMAKE_COMMAND} --build <BINARY_DIR>
+          --target seastar
+          --target seastar_testing
+          --target seastar_perf_testing
+          --target app_iotune
+        BUILD_BYPRODUCTS
+          <BINARY_DIR>/libseastar.$<IF:$<CONFIG:Debug,Dev>,so,a>
+          <BINARY_DIR>/libseastar_testing.$<IF:$<CONFIG:Debug,Dev>,so,a>
+          <BINARY_DIR>/libseastar_perf_testing.$<IF:$<CONFIG:Debug,Dev>,so,a>
+          <BINARY_DIR>/apps/iotune/iotune
+          <BINARY_DIR>/gen/include/seastar/http/chunk_parsers.hh
+          <BINARY_DIR>/gen/include/seastar/http/request_parser.hh
+          <BINARY_DIR>/gen/include/seastar/http/response_parser.hh
+        INSTALL_COMMAND "")
+    add_dependencies(Seastar::seastar Seastar)
+    add_dependencies(Seastar::seastar_testing Seastar)
+else()
```

Comment:
How does the difference between multi-mode and single-mode manifests?
And how does one do a single-mode build? By passing `--mode=dev`?

Reply:
> How does the difference between multi-mode and single-mode manifests? 

if user uses multi-config generator, it's required to configure seastar beforehand for all enabled modes. if user uses non-multi-config generator, it's not required to configure seastar beforehand. currently, `configure.py --use-cmake` always uses multi-config generator -- "Ninja Multi-Config". so that we can build any given build with a single `build.ninja` rule. 

> And how does one do a single-mode build? By passing `--mode=dev`?

one would need to run `cmake` manually. `cmake --use-cmake --mode=dev` passes `-G "Ninja Multi-Config"` to `cmake` . but if you think it's desirable to differentiate the two cases where we pass

- multiple modes to `configure.py` and 
- single modes to `configure.py`

i'd be happy to implement this. i think developer might want this behavior because it does offer better developer experience.

Reply:
> > How does the difference between multi-mode and single-mode manifests?
> 
> if user uses multi-config generator, it's required to configure seastar beforehand for all enabled modes. if user uses non-multi-config generator, it's not required to configure seastar beforehand. currently, `configure.py --use-cmake` always uses multi-config generator -- "Ninja Multi-Config". so that we can build any given build with a single `build.ninja` rule.

My question was for how seastar is included here: the true branch has add external project, the else branch has something else. What is the practical difference?

> 
> > And how does one do a single-mode build? By passing `--mode=dev`?
> 
> one would need to run `cmake` manually. `cmake --use-cmake --mode=dev` passes `-G "Ninja Multi-Config"` to `cmake` . but if you think it's desirable to differentiate the two cases where we pass
> 
>     * multiple modes to `configure.py` and
> 
>     * single modes to `configure.py`
> 
> 
> i'd be happy to implement this. i think developer might want this behavior because it does offer better developer experience.

I'm not sure this is needed. I usually invoke `configure.py` with all modes.

Reply:
> > > How does the difference between multi-mode and single-mode manifests?
> > 
> > 
> > if user uses multi-config generator, it's required to configure seastar beforehand for all enabled modes. if user uses non-multi-config generator, it's not required to configure seastar beforehand. currently, `configure.py --use-cmake` always uses multi-config generator -- "Ninja Multi-Config". so that we can build any given build with a single `build.ninja` rule.
> 
> My question was for how seastar is included here: the true branch has add external project, the else branch has something else. What is the practical difference? i actually answered your question already. 

but let me explain the intended use case here. i am quoting part of the the commit message:
```
- preserve the existing machinery of including Seastar only when
  building without multi-config generator. this allows users who don't
  use mult-config generator to build Seastar in-the-tree. the typical
  use case is the CI workflows performing the static analysis.
```

> 
> > > And how does one do a single-mode build? By passing `--mode=dev`?
> 
> I'm not sure this is needed. I usually invoke `configure.py` with all modes.

okay. i will add this option upon requests.